kernel/scheduler_epoch/src/lib.rs

```diff
@@ -28,92 +36,86 @@ const INITIAL_TOKENS: usize = 10;
-            priority_task_ref.tokens_remaining = tokens;
-            let task_ref = priority_task_ref.task.clone();
-            self.queue.push_back(priority_task_ref);
-            Some(task_ref)
-        } else {
-            None
+    fn try_next(&mut self) -> Option<TaskRef> {
+        while let Some(task) = self.have_tokens.pop_front() {
+            if task.task.is_runnable() {
+                if let Some(task) = self.add_epoch_task(task) {
+                    return Some(task);
+                }
+            } else {
+                task.task
+                    .expose_is_on_run_queue()
+                    .store(false, Ordering::Release);
+                // This check prevents an interleaving where `TaskRef::unblock` wouldn't add
+                // the task back onto the run queue. `TaskRef::unblock` sets the run state and
+                // then checks `is_on_run_queue` so we have to do the inverse.
+                if unlikely(task.task.is_runnable()) {
+                    if let Some(task) = self.add_epoch_task(task) {
+                        return Some(task);
+                    }
+                }
+            }
         }
+        None
     }
 
-    fn try_next(&mut self) -> Option<TaskRef> {
-        if let Some((task_index, _)) = self
-            .queue
-            .iter()
-            .enumerate()
-            .find(|(_, task)| task.is_runnable() && task.tokens_remaining > 0)
-        {
-            let chosen_task = self.queue.get(task_index).unwrap();
-            let modified_tokens = chosen_task.tokens_remaining.saturating_sub(1);
-            self.update_and_move_to_end(task_index, modified_tokens)
-        } else {
-            None
+    fn add_epoch_task(&mut self, mut task: EpochTaskRef) -> Option<TaskRef> {
+        task.task
+            .expose_is_on_run_queue()
+            .store(true, Ordering::Release);
+        match task.tokens_remaining.checked_sub(1) {
+            Some(new_tokens) => {
+                task.tokens_remaining = new_tokens;
+                let task_ref = task.task.clone();
+                self.have_tokens.push_back(task);
+                Some(task_ref)
+            }
+            None => {
+                self.out_of_tokens.push(task);
+                None
+            }
         }
     }
 
     fn assign_tokens(&mut self) {
-        // We begin with total priorities = 1 to avoid division by zero
-        let mut total_priorities: usize = 1;
-
-        // This loop calculates the total priorities of the runqueue
-        for (_i, t) in self.queue.iter().enumerate() {
-            // we assign tokens only to runnable tasks
-            if !t.is_runnable() {
-                continue;
-            }
-
-            total_priorities = total_priorities
-                .saturating_add(1)
-                .saturating_add(t.priority as usize);
+        while let Some(task) = self.out_of_tokens.pop() {
+            self.have_tokens.push_back(task);
```

Comment:
nit: Is there a way to do this in one batch rather than item-by-item? I guess not when we're using `VecDeque` for one list and `Vec` for the other...?